### ISSUE TITLE
SpawnEntity, DeleteEntity, SetPose Services

### ISSUE THREAD

**Issue by retinfai** [gazebosim/ros_gz#363]
## Desired behavior
ros_gz_bridge support SpawnEntity, DeleteEntity, and SetPose services, so I can access them through ROS2 client calls.

## Alternatives considered
Running gz services through Subprocesses

## Implementation suggestion
Is it just `ros_gz_bridge/src/service_factories/ros_gz_interfaces.cpp` that needs to be changed? Not too sure, but just a thought.

## Additional context
Gazebo Garden, ROS2 Humble.
The interfaces for the services, request types, and response types are already supported in `ros_gz_interfaces` or other

**Comment by azeey** [comment]
This would be a nice addition. A PR would be welcome 😉 .

**Comment by retinfai** [comment]
I can give it a good go. First open source contribution attempt aha

**Comment by zp-yang** [comment]
How is the progress on implementing these services? I would also like to help if needed.

**Comment by retinfai** [comment]
I've begun work on this fork and hacked a solution for my own purposes here: https://github.com/retinfai/ros_gz

But more tweaking is needed before I want to submit for review.

Would love collaboration – I haven't found a lot of time recently.

**Comment by zp-yang** [comment]
which ros2 version and gazebo version are you using? I am running humble and garden.

**Comment by retinfai** [comment]
I'm using that too.

**ROS2 Humble** & **Gazebo Garden**

**Comment by zp-yang** [comment]
Cool, what do you think still needs to be done? You seemed to have all of the features working. I am mainly interested in the set_pose service and it works on my side as well.

**Comment by retinfai** [comment]
It's more like it's incomplete. I don't convert every field. 

But for set pose specifically, it gets confusing for me when gz.msgs.Pose doesn't map one to one to anything in the ROS side, so want to know the best practice for mapping 1 to N and vice versa. Set pose request (Entity and Pose on the ROS side) map to gazebo Pose but not all fields in Entity and Pose are used, so I exclude them. That sort of thing

I'm glad it works on your side. Now that I think about it, you're right: just need to iron out some things

**Comment by Mildred34** [comment]
I was wondering how to set the location of an entity spawned within Gazebo Garden. 
But it's not officially supported.

**Comment by azeey** [comment]
> I was wondering how to set the location of an entity spawned within Gazebo Garden. But it's not officially supported.

This is definitely supported. Checkout the message definition https://github.com/gazebosim/gz-msgs/blob/gz-msgs9/proto/gz/msgs/entity_factory.proto#L65

**Comment by Mildred34** [comment]
@azeey Looks like during spawn only by reading at the comment. 
I was thinking about setting (or resetting) the location afterwards.

**Comment by retinfai** [comment]
> @azeey Looks like during spawn only by reading at the comment. 
> 
> I was thinking about setting (or resetting) the location afterwards.
> 
> 
> 
> 
If you're talking about the bridge supporting you're correct.

But if you're talking about gazebo supporting it, gazebo does have a /model/<name>/set_pose service or something similar. You can call from a process

**Comment by Mildred34** [comment]
@retinfai I'm not an expert but to call a gazebo service from a ros node, I need the bridge support ?
If not, I'm going to investigate inside the set_pose service, you mentionned !

**Comment by retinfai** [comment]
> @retinfai I'm not an expert but to call a gazebo service from a ros node, I need the bridge support ?
> 
> If not, I'm going to investigate inside the set_pose service, you mentionned !

That's correct, calling the service from a Ros node does require bridge support . Just wanted to clarify. 

In saying that, you can pull the fork, which does support it, while I fix issues and create a PR

**Comment by Mildred34** [comment]
@retinfai Yes I'll do that! Thanks for the help

**Comment by Trana** [comment]
I tested the create/remove/set_pose services on Jazzy and Harmonic and the changes in this PR worked fine for me.

**Comment by j-rivero** [comment]
:information_source:  The Gazebo core team has included the review of PR into a community effort to attract new contributions, [see the message in LinkdIn](https://www.linkedin.com/posts/open-source-robotics-foundation_qos-docs-not-being-generated-in-sphinx-activity-7304943865529057281-7wFO?utm_source=share&utm_medium=member_desktop&rcm=ACoAABJBGkgBZvpxU2P6e4d-EBpI5IkiXVtoUVQ). I wrote some context and introduction about how to help in https://gist.github.com/j-rivero/9ea1abe2441c8dbf3cc4dffccfbdaff8.

**Comment by PerFrivik** [comment]
does this work in jazzy with harmonic?

**Comment by elmeripk** [comment]
@PerFrivik Works correctly for me with jazzy+harmonic. Just be sure to update your packages.
